htmresearch/algorithms/apical_modulation_temporal_memory.py

```diff
@@ -0,0 +1,917 @@
+      Matching apical segments selected for learning in bursting columns
+
+    - apicalSegmentsToPunish (numpy array)
+      Apical segments that should be punished for predicting an inactive column
+
+    - newApicalSegmentCells (numpy array)
+      Cells in bursting columns that were selected to grow new apical segments
+    """
+
+    # Cells with active apical segments
+    learningActiveApicalSegments = self.apicalConnections.filterSegmentsByCell(
+      activeApicalSegments, learningCells)
+
+    # Cells with matching apical segments
+    learningCellsWithoutActiveApical = np.setdiff1d(
+      learningCells,
+      self.apicalConnections.mapSegmentsToCells(learningActiveApicalSegments))
+    cellsForMatchingApical = self.apicalConnections.mapSegmentsToCells(
+      matchingApicalSegments)
+    learningCellsWithMatchingApical = np.intersect1d(
+      learningCellsWithoutActiveApical, cellsForMatchingApical)
+    learningMatchingApicalSegments = self._chooseBestSegmentPerCell(
+      self.apicalConnections, learningCellsWithMatchingApical,
+      matchingApicalSegments, apicalPotentialOverlaps)
+
+    # Cells that need to grow an apical segment
+    newApicalSegmentCells = np.setdiff1d(learningCellsWithoutActiveApical,
+                                         learningCellsWithMatchingApical)
+
+    # Incorrectly predicted columns
+    correctMatchingApicalMask = np.in1d(
+      cellsForMatchingApical / self.cellsPerColumn, activeColumns)
+
+    apicalSegmentsToPunish = matchingApicalSegments[~correctMatchingApicalMask]
+
+    return (learningActiveApicalSegments,
+            learningMatchingApicalSegments,
+            apicalSegmentsToPunish,
+            newApicalSegmentCells)
+
+
+  @staticmethod
+  def _calculateSegmentActivity(connections, activeInput, connectedPermanence,
+                                activationThreshold, minThreshold, apicallyActiveCells=()):
+    """
+    Calculate the active and matching segments for this timestep.
+
+    @param connections (SparseMatrixConnections)
+    @param activeInput (numpy array)
+
+    @return (tuple)
+    - activeSegments (numpy array)
+      Dendrite segments with enough active connected synapses to cause a
+      dendritic spike
+
+    - matchingSegments (numpy array)
+      Dendrite segments with enough active potential synapses to be selected for
+      learning in a bursting column
+
+    - potentialOverlaps (numpy array)
+      The number of active potential synapses for each segment.
+      Includes counts for active, matching, and nonmatching segments.
+    """
+
+    # Active apical segments lower the activation threshold for basal (lateral) segments
+    overlaps = connections.computeActivity(activeInput, connectedPermanence)
+    outrightActiveSegments = np.flatnonzero(overlaps >= activationThreshold)
+    conditionallyActiveSegments = np.flatnonzero((overlaps < activationThreshold) & (overlaps >= activationThreshold * 0.75))
```

Comment:
I agree that this should be a threshold parameter, not a factor. My vote is that we add a new parameter, not use minThreshold. minThreshold is essentially a "learningThreshold" and probably shouldn't be tangled with this.